jaeger-core/src/main/java/io/jaegertracing/internal/clock/SystemClock.java

```diff
@@ -14,19 +14,45 @@
 
 package io.jaegertracing.internal.clock;
 
+import lombok.val;
+
 /**
- * Default implementation of a clock that delegates its calls to the system clock. The
- * microsecond-precision time is simulated by (millis * 1000), therefore the
- * {@link #isMicrosAccurate()} is false.
+ * This implementation of the system-clock will provide true microseconds accurate timestamp,
+ * given that the JVM supports it (JDK 9 and above).
+ * <p>
+ * The actual timestamp generation implementation for both scenarios can be found in {@link CurrentTimeSupplier}.
  *
- * @see System#currentTimeMillis()
+ * @author <a href="mailto:ishinberg0@gmail.com">Idan Sheinberg</a>
+ * @see CurrentTimeSupplier#micros()
  * @see System#nanoTime()
  */
 public class SystemClock implements Clock {
 
+  private static final boolean MICROS_ACCURATE;
+
+  private static final CurrentTimeSupplier CURRENT_TIME_SUPPLIER;
+
+  private static int getJavaVersion() {
+    val sections = System.getProperty("java.version").split("\\.");
+    val major = Integer.parseInt(sections[0]);
+    return major == 1 ? Integer.parseInt(sections[1]) : major;
+  }
+
+  private static CurrentTimeSupplier getCurrentTimeSupplier() {
+    return MICROS_ACCURATE
+        ? CurrentTimeSupplier.MicrosAccuracy.INSTANCE
+        : CurrentTimeSupplier.MillisAccuracy.INSTANCE;
+  }
+
+  static {
+    val version = getJavaVersion();
+    MICROS_ACCURATE = version >= 9;
+    CURRENT_TIME_SUPPLIER = getCurrentTimeSupplier();
+  }
+
   @Override
   public long currentTimeMicros() {
-    return System.currentTimeMillis() * 1000;
+    return CURRENT_TIME_SUPPLIER.micros();
```

Comment:
```java
Instant now = Instant.now();
return now.atZone(ZoneId.systemDefault()).toEpochSecond() * 1000000 + now.getNano() / 1000;
```
Both JDK8 and JDK9+ are supported.

Reply:
The code you presented is wrong.
`getNano()` returns a monotonous clock value, not an RTC clock.
microseconds (maximaul) resolution is supported only from JDK 9 and above.

You can see additional information in the original issue https://github.com/jaegertracing/jaeger-client-java/issues/710 and also search on stack overflow.

Closing this discussion